src/bsoncxx/builder/list/array.hpp

```diff
@@ -0,0 +1,195 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/config/prelude.hpp>
+
+#include <bsoncxx/builder/core.hpp>
+#include <bsoncxx/types/bson_value/value.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace builder {
+namespace list {
+using namespace bsoncxx::types;
+
+///
+/// A JSON-like builder for creating BSON arrays.
+///
+class array {
+    template <bool B, class T = void>
+    using enable_if_t = typename std::enable_if<B, T>::type;
+
+    template <class T>
+    using decay_t = typename std::decay<T>::type;
+
+   public:
+    ///
+    /// Default constructor. Creates an empty array builder.
+    ///
+    array() = default;
+
+    ///
+    /// Converting constructor. Initializes each element of the array builder with the corresponding
+    /// value in std::forward<Args>(args).
+    ///
+    /// @param args values used to initialize each element of the array builder
+    ///
+    /// @note a bsoncxx::types::bson_value::value is direct-initialized from each argument.
+    ///
+    /// @see bsoncxx::types::bson_value::value
+    ///
+    template <typename... Args>
+    array(Args&&... args) {
+        this->_append(_core, std::forward<Args>(args)...);
+    }
+
+    ///
+    /// Move constructor. Constructs the array builder with the contents of other using move
+    /// semantics.
+    ///
+    /// @param other another array builder to use as source to initialize this array builder with
+    ///
+    array(array&& other) noexcept = default;
+
+    ///
+    /// Replaces the contents with those of other using move semantics.
+    ///
+    /// @param other another array builder to use as source to initialize this array builder with
+    ///
+    array& operator=(array&& other) noexcept = default;
+
+    ///
+    /// Returns an owning bsoncxx::array::value.
+    ///
+    /// @return An owning array::value representing the entire contents of the array builder.
+    ///
+    /// @see bsoncxx::array::value
+    ///
+    operator bsoncxx::array::value() {
+        return this->extract();
+    }
+
+    ///
+    /// Returns an owning bsoncxx::array::value.
+    ///
+    /// @return An owning array::value representing the entire contents of the array builder.
+    ///
+    /// @see bsoncxx::array::value
+    ///
+    bsoncxx::array::value extract() {
+        return _core.extract_array();
+    }
+
+    ///
+    /// Appends array builder rhs.
+    ///
+    /// @param rhs array builder to append
+    ///
+    /// @return *this
+    ///
+    array& operator+=(array&& rhs) {
+        this->append(std::move(rhs));
+        return *this;
+    }
+
+    ///
+    /// Appends array builder rhs.
+    ///
+    /// @param rhs array builder to append
+    ///
+    /// @return *this
+    ///
+    array& append(array&& rhs) {
```

Comment:
Extracting `rhs` leaves the source builder in an empty/invalid state.

```
    builder::list::array a = {"a", 1};
    bsoncxx::document::value d = builder::list::document{{"x", a}, {"y", a}};
    std::cout << bsoncxx::to_json (d.view()) << std::endl;
    // { "x" : [ "a", 1 ], "y" : [  ] }
```

That seems like very surprising behavior to me.

Should we copy instead?

Reply:
Both `list::array` and `list::document` wrap the core builder which has no copy operators. The alternative solution is to do this 
```
        builder::list::array a = {"a", 1};
        auto arr = a.extract();

        bsoncxx::document::value d = builder::list::document{{"x", arr}, {"y", arr}};
        std::cout << bsoncxx::to_json (d.view()) << std::endl;
        // { "x" : [ "a", 1 ], "y" : [ "a", 1 ] }
```
I'll make a note in the implicit conversion operators.

Reply:
Edit: Also, @samantharitter brought up that the implicit conversion operator should be removed entirely since it isn't clear when it's used and subsequently when the builder is no longer valid. That approach seems like the best one. 

Since we can't copy the core builder, it'd make sense to remove this method and other methods that call extract under the hood to avoid confusion. That'll force the user to call `extract` when they'd like to use the array or document builders as BSON arrays or documents, respectively.

Reply:
I agree with removing the implicit conversion and API that calls `extract` under the hood to avoid surprises in API.

Additionally, requiring callers to call `extract` does not seem much more verbose for one-liners. Right now, you need to cast / define the type:
```
document::value doc = list::document({ "x", 1 })
```

With `extract`, callers can use `auto`.

```
auto doc = list::document({ "x", 1 }).extract()
```

Reply:
Yep. The only time `extract` is a bit strange is during an example like:
```
list::array arr = {"foo", list::array{"bar", "baz"}.extract()};
```
In other words, nested document and array builders (even temporary ones) need to be `extract`ed. Maybe there's some template magic that could fix this, but unless you think that's a big problem, we'll require that syntax for nested docs and arrays for now.

Reply:
I think this is fine, and it's consistent with the existing builders, so users who have used the driver in the past won't be surprised by it.  The name `extract()` implies that you are stealing something out from the object, so it shouldn't be too surprising that it leaves the builder in a moved-from-like state.

Reply:
There's a trade-off between convenience and safety for requiring `extract` or doing it implicitly in the list builder. I don't have a suggestion for anything that solves both, but I lean towards safety.

I think we could add implicit conversions later in a safe way if we choose to. Perhaps we could update the core builder to permit copying, or access the core builder's underlying bson data and append it via libbson API directly.